### PR TITLE
Fix PSA crypto inconsistencies in agreement+derivation

### LINKED ISSUE #3741: Clarification on PSA key algorithm restriction for agreement/derivation
### Description
- Type: Inconsistency/Question

---------------------------------------------------------------
## Question
PSA says (on the topic of key usage restrictions) :
`
In addition to the usage bitmask, a policy specifies which algorithm is permitted with the key. This specification only defines policies that restrict keys to a single algorithm, which is consistent with both common practice and security good practice.
`

This is implemented super strict inside of `psa_crypto.c`, and I'm wondering why it does so. Consider a key used for ECDH, stored non-exportable persistent. The algorithm declared when generating the key needs to match strictly with the algorithm the key is used for.

1. If one wants to use the key with `psa_raw_key_agreement()`, it needs to be defined with `PSA_ALG_ECDH`, and the core doesn't care what happens with the shared secret.
2. If one wants to use the key with `psa_key_derivation_key_agreement()`, then it needs to be defined with `PSA_ALG_KEY_AGREEMENT(PSA_ALG_ECDH, PSA_ALG_xKDF(PSA_ALG_xHASH))`, all depending on which KDF is used. The key is thus bound to one specific combination of both agreement, derivation and hash primitive algorithms.

The question then begs: a key declared for use with `PSA_ALG_ECDH`, shouldn't that be allowed to be presented to `psa_key_derivation_key_agreement` for use of a combined agreement/derivation operation? Since it is under-specified in a way?

I assume this'll be a larger problem when porting the TLS portion on top of PSA Crypto with long-term keys. If you have an ECDH key, and didn't specify a restriction on derivation algorithm, it really shouldn't matter for the usage restriction which derivation algorithm you're using after the agreement step.

### LINKED ISSUE #3742: Cannot derive PSA key directly through key agreement-and-derivation
### Description
- Type: Bug

---------------------------------------------------------------
## Bug

**OS**  
N/A

**mbed TLS build:**  
Development

**Peer device TLS stack and version**  
N/A

**Expected behavior**   
Using PSA crypto to streamline the operation of agreeing and deriving a key should work. E.g.:
```
psa_key_derivation_setup( &hkdf_ctx, PSA_ALG_KEY_AGREEMENT(PSA_ALG_ECDH, PSA_ALG_HKDF(PSA_ALG_SHA_256)) );
psa_key_derivation_set_capacity( &hkdf_ctx, PSA_BITS_TO_BYTES(256) );
psa_key_derivation_input_bytes( &hkdf_ctx, PSA_KEY_DERIVATION_INPUT_SALT, salt, sizeof(salt) );
psa_key_derivation_input_bytes( &hkdf_ctx, PSA_KEY_DERIVATION_INPUT_INFO, NULL, 0 );
psa_key_derivation_key_agreement( &hkdf_ctx, PSA_KEY_DERIVATION_INPUT_SECRET, ecdh_handle, counterpart, sizeof(counterpart) );
psa_key_derivation_output_key( &derived_key_attr, &hkdf_ctx, &derived_key_handle );
```

**Actual behavior**  
`operation->can_output_key` is not being set if a key is added to the derivation operation through `psa_key_derivation_key_agreement`, meaning `psa_key_derivation_output_key` will refuse to output a key if the derivation is key agreement based. It does that with `PSA_ERROR_NOT_PERMITTED`.

### PR DESCRIPTION
**Pull request by stevew817**
## Description
Fix #3741: Allow key agreement inside derivation with a key that's allowed for the relevant agreement.
Fix #3742: After input of a key as SECRET in the derivation, allow the derivation result to be used as key.

## Status
**READY**

## Requires Backporting
NO  

## Migrations
NO

## Additional comments
Is a changelog needed?

## Todos
- [X] Tests
- [x] Changelog updated

## Steps to test or reproduce
Unit tests.

### CODE REVIEW

**Comment by gilles-peskine-arm** [comment]
These are two separate changes. They're small enough to put in the same pull request, but please make separate commits.

Yes, please add changelog entries (one Bugfix and one Features).

**Comment by stevew817** [comment]
@gilles-peskine-arm split in two commits and added changelog entries. Ready for review.

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
Looks overall good to me. I have a few suggestions for improvement. I've checked that the unit tests that are supposed to fail when the corresponding changes to psa_crypto.c are reverted actually fail.

File: library/psa_crypto.c [hunk 1]
```diff
@@ -5877,6 +5877,11 @@ static psa_status_t psa_key_agreement_internal( psa_key_derivation_operation_t *
                                                 shared_secret,
                                                 shared_secret_length );
 
+    /* If a private key has been added as SECRET, we allow the derived
+     * key material to be used as a key in PSA Crypto. */
+    if( step == PSA_KEY_DERIVATION_INPUT_SECRET )
```
[the remark this answers is not in the document]
Comment (stevew817) [reply library/psa_crypto.c:5882]: Sure.

**Comment by stevew817** [comment]
@ronald-cron-arm all your feedback is now applied.

**Review by ronald-cron-arm — APPROVED** [review]
Thanks for the changes. It looks good to me now.

**Review by paul-elliott-arm — APPROVED** [review]
Looks good to me with latest changes.

**Pull request merged**